src/Waives.Http/RequestHandling/AccessTokenService.cs

```diff
@@ -0,0 +1,57 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Caching.Memory;
+using Polly;
+using Polly.Caching;
+using Polly.Caching.Memory;
+using Waives.Http.Responses;
+
+namespace Waives.Http.RequestHandling
+{
+    internal class AccessTokenService : IDisposable
+    {
+        private readonly string _clientId;
+        private readonly string _clientSecret;
+        private readonly IHttpRequestSender _requestSender;
+        private readonly MemoryCache _cache;
+        private readonly AsyncCachePolicy<AccessToken> _cachePolicy;
+
+        internal AccessTokenService(string clientId, string clientSecret, IHttpRequestSender requestSender)
+        {
+            _clientId = clientId;
+            _clientSecret = clientSecret;
+            _requestSender = requestSender;
+
+            _cache = new MemoryCache(new MemoryCacheOptions());
+            _cachePolicy = Policy.CacheAsync(
+                new MemoryCacheProvider(_cache).AsyncFor<AccessToken>(),
+                new ResultTtl<AccessToken>(t => new Ttl(t.LifeTime)),
```

Comment:
Good point. Allowing for requests timing out, the "max overall retry period" is in the region of 15-20 minutes, but the default 2-minute timeout period can be overridden by consuming applications. Maybe we need to factor that timeout period into the calculation here?

Reply:
The jwt should have an absolute expiry time encoded within it.

Reply:
Do we need a memory cache? wouldn't a single variable suffice?

Reply:
Yes, except we then have to explicitly consider things like thread-safety, enforcing the TTL, etc. We get all that for free using Polly and MemoryCache 🤷‍♂ 

The JWT does have an absolute expiry time, yes. I figured that the lifetime was a clean match for the TTL on the cache item, and absolved us of doing error-prone date maths.

Reply:
One reason we opted to use the absolute expiry time in surf was to allow us to cache the token in the webapi (which we don't yet do, tbf) - the `ExpiresAt` property only gives an approximation of the time remaining until it expires.

Reply:
aa17592 ensures the cache item TTL is 1hr less than the token lifetime. I don't see that whether or not the token is cached in the API has any effect here, as it will be cached locally and so the client won't make any token requests to the API until it's expired. 

Should we perhaps set `Cache-Control` headers on our token endpoint to encourage good behaviour by clients?